File: TOOLS/lua2string.py
#!/usr/bin/env python3

#
# This file is part of mpv.
#
# mpv is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# mpv is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
#

import sys
import string
import subprocess


def cstr(s):
    conv = ["\\%o" % c for c in range(256)]

    for c in string.ascii_letters:
        conv[ord(c)] = c

    return '"%s"' % "".join(conv[c] for c in s)


[_, lua_version, luac_prog, strip, input, output] = sys.argv

if luac_prog:
    # Precompile Lua to bytecode. We use `string.dump` for LuaJIT because it supports the "strip" parameter,
    # unlike Lua 5.1 and Lua 5.2 where we must use the CLI to allow stripping.
    #
    # We use `string.dump` directly for LuaJIT because "luajit -b" is
    # handled by "jit.bcsave" module that may not work if user has messed up
    # her `$LUA_PATH`. Also, checking availability of "-d" (the "d" flag,
    # deterministic code generation) would require some extra code.
    #
    # `string.dump()`
    lua_chunk = subprocess.check_output(
        [
            luac_prog,
            "-e",
            "assert(io.write(string.dump(assert(loadfile(arg[1])), assert(arg[2]))))",
            # We read an empty script from stdin just to be able pass some
            # arguments to "-e".
            "-",
            input,
            ("s" if strip else "") + "d",
        ]
        if lua_version == "luajit"
        else [
            luac_prog,
            *(["-s"] if strip else []),
            "-o",
            "-",
            input,
        ]
    )
else:
    # Pass through plain text Lua.
    with open(input, "rb") as f:
        lua_chunk = f.read()

with open(output, "w") as f:
    f.write(
        f"""\
{{
    .start = {cstr(lua_chunk)},
    .len = {len(lua_chunk)},
}}
"""
    )

File: TOOLS/lua2string.lua
--[[
This file is part of mpv.

mpv is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

mpv is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
]]

local lua_version, luac, luac_strip, luac_strip_cmd, input, output = ...
luac = luac == "yes"
luac_strip = luac_strip == "yes"

local function to_cstr(s)
    return ('"%s"'):format(s:gsub("%A", function(c)
        return ("\\%o"):format((c):byte())
    end))
end

local function get_lua_chunk()
    if not luac then
        local f = assert(io.open(assert(input)))
        return assert(f:read("*a"))
    end

    -- Lua <5.3 does not have a "strip" parameter on `string.dump` so we must
    -- use an external command for it.
    if luac_strip and luac_strip_cmd ~= "" then
        local f = assert(io.popen(luac_strip_cmd:format(input)))
        local s = assert(f:read("*a"))
        assert(s ~= "", ("%q returned no output"):format(luac_strip_cmd))
        return s
    end

    local fn = assert(loadfile(assert(input)))

    -- Request deterministic bytecode from LuaJIT (if supports).
    if lua_version == "luajit" then
        return string.dump(fn, luac_strip and "sd" or "d")
    end

    return string.dump(fn, luac_strip)
end

local lua_chunk = get_lua_chunk()

local c_code = ([[
{
    .start = %s,
    .len = %d,
}
]]):format(to_cstr(lua_chunk), #lua_chunk)

local outfile = assert(io.open(assert(output), "wb"))
assert(outfile:write(c_code))
assert(outfile:close())
